src/lib.rs

```diff
@@ -42,456 +42,417 @@
+            (mapped
+                .parse::<u8>()
+                .expect("color tables should contain valid numbers")
+                + 10)
+                .to_string()
+        } else {
+            mapped.to_string()
+        };
 
-            if bg && self.depth <= Depth::Low {
-                value = (value.parse::<u8>().ok()? + 10).to_string()
-            };
+        fill_template(templater(self.depth), &value)
+    }
 
-            Some(fill_template(template, &value))
-        }
+    /// Writes the prompt to stdout, then reads a line from input,
+    /// and returns it (excluding the trailing newline).
+    pub fn input(&self, prompt: &str) -> std::io::Result<String> {
+        print!("{}", self.convert(prompt));
+        stdout().flush()?;
+
+        let mut inp = String::new();
+        stdin().read_line(&mut inp)?;
+        Ok(inp.trim_end().to_owned())
     }
 
     /// Resets the formatting back to the default.
-    pub fn reset(&self) {
-        print!("{}", self.convert(format!("{}r", self.marker)));
+    pub fn reset() {
+        print!("\x1b[0m");
+    }
+
+    /// Clears the current line
+    pub fn clear_line() {
+        print!("\x1b[2K");
+    }
+
+    /// Clears the screen
+    pub fn clear_screen() {
+        print!("\x1b[2J");
+    }
+
+    /// Escape all format markers in a string
+    ///
+    /// # Example
+    /// ```rust
+    /// # use dahlia::Dahlia;
+    /// let d = Dahlia::default();
+    /// let str = d.escape("&aHello &cWorld");
+    /// assert_eq!(str, "&_aHello &_cWorld");
+    /// ```
+    pub fn escape(&self, str: &str) -> String {
+        str.replace(self.marker, self.patterns.escaped())
     }
 
     /// Returns a string with all the possible formatting options.
     pub fn test(&self) -> String {
         self.convert(
-            "0123456789abcdefg"
+            &"0123456789abcdef"
                 .chars()
                 .map(|ch| format!("{m}{ch}{ch}", m = self.marker))
                 .chain(
-                    "lmno"
+                    "hijklmno"
                         .chars()
-                        .map(|ch| format!("{m}r{m}{ch}{ch}", m = self.marker)),
+                        .map(|ch| format!("{m}R{m}{ch}{ch}", m = self.marker)),
                 )
                 .collect::<String>(),
         )
-        .to_string()
+        .into_owned()
     }
 }
 
-fn re(string: &str) -> Regex {
-    Regex::new(string).unwrap()
+fn hex_to_ansi(hex: &str, templater: fn(Depth) -> &'static str) -> String {
+    let hex_digits = hex
+        .chars()
+        .map(|ch| ch.to_digit(16))
+        .collect::<Option<Vec<_>>>()
+        .expect("the regex should only match valid hexadecimal digits");
+
+    let [r, g, b] = match &hex_digits[..] {
+        // if there are only 3 digits, "duplicate" each
+        [r, g, b] => [r, g, b].map(|&d| (0x11 * d).to_string()),
+        [r1, r2, g1, g2, b1, b2] => {
+            [(r1, r2), (g1, g2), (b1, b2)].map(|(h, l)| (h * 0x10 + l).to_string())
```

Comment:
Do `h` and `l` mean anything in particular?

Reply:
high/low half of the byte